bokehjs/src/lib/core/util/templating.ts

```diff
@@ -3,84 +3,146 @@ import * as Numbro from "numbro"
 import tz = require("timezone")
 
 import {escape} from "./string"
-import {isNumber, isString} from "./types"
+import {isNumber, isString, isArray, isTypedArray} from "./types"
 
 import {ColumnarDataSource} from "models/sources/columnar_data_source"
+import {ImageIndex} from "../../models/glyphs/image"
 import {CustomJSHover} from 'models/tools/inspectors/customjs_hover'
 
-function _format_number(num: string | number): string {
-  if (isNumber(num)) {
+export type FormatterType = "numeral" | "printf" | "datetime"
+export type FormatterSpec = CustomJSHover | FormatterType
+export type Formatters = {[key: string]: FormatterSpec} | null
+export type FormatterFunc = (value:any, format:string, special_vars: Vars) => string
+export type Index = number | ImageIndex
+export type Vars = {[key: string]: any}
+
+export const DEFAULT_FORMATTERS = {
+  "numeral"  : function(value:any, format:string, special_vars: Vars) { special_vars; return Numbro.format(value, format) },
+  "datetime" : function(value:any, format:string, special_vars: Vars) { special_vars; return tz(value, format)            },
+  "printf"   : function(value:any, format:string, special_vars: Vars) { special_vars; return sprintf(format, value)       },
```

Comment:
Precede an argument name with `_`, so `_special_vars`. This will silence the compiler regarding unused function arguments. Note this apply only to function arguments.